File: crates/transport-cli/src/commands/obs.rs
//! `spiritstream-cli obs …` — OBS Studio WebSocket integration (runtime only).
//!
//! Connection lifecycle + stream control. OBS *settings* are owned by the active
//! profile (the single source of truth): edit them with
//! `profile set <name> --set obs.host=… --set obs.autoConnect=…` and read them
//! with `profile show <name>`. There is no `obs config` / `obs set-config` —
//! those mutated an ephemeral per-invocation handler that never persisted.

use clap::Subcommand;
use spiritstream_core::ServiceRegistry;

use crate::error::CliError;
use crate::output::Output;

#[derive(Debug, Subcommand)]
pub enum ObsCmd {
    /// Snapshot of OBS connection + streaming state.
    State,
    /// Connect to the OBS WebSocket using the active profile's settings.
    Connect,
    /// Disconnect from OBS.
    Disconnect,
    /// Start streaming in OBS (requires an active connection).
    StreamStart,
    /// Stop streaming in OBS.
    StreamStop,
}

pub async fn run(
    cmd: ObsCmd,
    registry: &ServiceRegistry,
    out: &mut Output,
) -> Result<(), CliError> {
    match cmd {
        ObsCmd::State => {
            let state = registry.obs.get_state().await;
            out.emit(&state)?;
            Ok(())
        }
        ObsCmd::Connect => {
            registry.obs.connect(registry.events.clone()).await?;
            out.emit(&serde_json::json!({ "connected": true }))?;
            Ok(())
        }
        ObsCmd::Disconnect => {
            registry.obs.disconnect(registry.events.clone()).await?;
            out.emit(&serde_json::json!({ "disconnected": true }))?;
            Ok(())
        }
        ObsCmd::StreamStart => {
            registry.obs.start_stream().await?;
            out.emit(&serde_json::json!({ "started": true }))?;
            Ok(())
        }
        ObsCmd::StreamStop => {
            registry.obs.stop_stream().await?;
            out.emit(&serde_json::json!({ "stopped": true }))?;
            Ok(())
        }
    }
}

File: packages/api-client/src/api/obs.ts
import type { ObsState } from '@spiritstream/types';
import {
  v1ObsStateProxy,
  v1ObsConnectProxy,
  v1ObsDisconnectProxy,
  v1ObsStartStreamProxy,
  v1ObsStopStreamProxy,
  v1ObsIsConnectedProxy,
} from '../generated';

// OBS settings are NOT read/written through the API. The active profile is the
// single source of truth: settings persist via the profile save endpoint, and
// the backend syncs its runtime handler from the profile on activation/save.
// This client only drives the runtime connection (connect/disconnect/stream).

export const obs = {
  getState: async (): Promise<ObsState> => {
    const { data } = await v1ObsStateProxy({ throwOnError: true });
    return data as ObsState;
  },
  connect: async (): Promise<void> => {
    await v1ObsConnectProxy({ throwOnError: true });
  },
  disconnect: async (): Promise<void> => {
    await v1ObsDisconnectProxy({ throwOnError: true });
  },
  startStream: async (): Promise<void> => {
    await v1ObsStartStreamProxy({ throwOnError: true });
  },
  stopStream: async (): Promise<void> => {
    await v1ObsStopStreamProxy({ throwOnError: true });
  },
  isConnected: async (): Promise<boolean> => {
    const { data } = await v1ObsIsConnectedProxy({ throwOnError: true });
    return data.connected;
  },
};
